### ISSUE TITLE
Solution for replication with large lag of data

### ISSUE THREAD

**Issue by amyangfei** [pingcap/tiflow#1150]
## Feature Request

**Is your feature request related to a problem? Please describe:**

In some extreme scenarios, TiCDC could meet a large replication lag, and the data needs to be caught up is too large for TiCDC to handle. The scenarios include:
1. The downstream is slow, and a lot of data is accumulated in TiCDC
2. The replication task is paused for a long time, such as 3 days, and the user wants to continue the task from the existing checkpoint.

We have brainstorm ideas for both scenario-1 and scenario-2, including

- scenario-1
  - Add flow control in `EventFeed` API from TiKV to TiCDC. This may solve the OOM in TiCDC but could cause OOM in TiKV
  - Use disk to buffer data before writing to sink https://github.com/pingcap/ticdc/issues/1096
- scenario-2
  - Provide a powerful unified sorter to handle incremental scan with a large data set. https://github.com/pingcap/ticdc/pull/972, https://github.com/pingcap/ticdc/issues/1148. This solution still has an incremental scan data size limitation.
  - Use BR incremental backup and restore as a workaround. This solution can't be used when replicating to non-TiDB downstream.
  - Support temporary output sorted data to file. This scenario happens when upstream has write traffic all the time and full back up and restore could cost long. In this scenario we can start TiCDC before full dump/export and output sorted change data into file.
  - TiKV provides a staged incremental scan API, such as when TiCDC wants to scan incremental data with tso `[1000, 4000)`, TiKV can separate the scan interval into several parts, supposing 3 here, then TiKV returns the incremental data with tso `[1000, 2000)`, `[2000, 3000)`, `[3000, 4000)` respectively. 

**Describe the feature you'd like:**

- We must define the scale of data that TiCDC can support, and select different solutions for each scale. (While from the point of users, it is difficult to find out how much incremental data there exists)
- Make development plans and task breakdown for the selected solutions

**TODO List**

- Bottom line
  - [x] Double check when creating a new changefeed with replication lag larger than **3 days**.
  - [ ] Degrade the service (such as stop the changefeed) when the incremental data is larger than threshold.
  - [ ] Provide document to guide user how to recover data flow when replication lag is out of the ability of TiCDC.
- Solve the OOM issue in a defined scale
  - [x] Make unified sorter available
  - [x] Solve the OOM issue in TiKV when TiCDC receives messages slower than TiKV sends. https://github.com/tikv/tikv/pull/9885
  - [x] Solve the OOM in TiCDC when sink is slower than the sorter. https://github.com/pingcap/ticdc/pull/1566

**Comment by glkappe** [comment]
Wow, when can this requirement be initiated?
Recently, there are many similar problems in asktug, but there is no solution.

BR's incremental backup and recovery cannot meet the recovery scenarios of online clusters, which is really tricky.

Usability problem is big!!!!

https://asktug.com/t/topic/63728  https://asktug.com/t/topic/63770  https://asktug.com/t/topic/37391 
 https://asktug.com/t/topic/63754

**Comment by glkappe** [comment]
@overvenus any update??

**Comment by amyangfei** [comment]
It is far from an usability problem, the solution should be carefully designed and it will be a large feature. @glkappe

**Comment by IANTHEREAL** [comment]
`CDC:ErrBufferReachLimit]puller mem buffer reach size limit"` means that there is no request to access TiCDC invalid GC data, but TiCDC cannot be replcated. This is indeed a possible problem, right?

**Comment by amyangfei** [comment]
> `CDC:ErrBufferReachLimit]puller mem buffer reach size limit"` means that there is no request to access TiCDC invalid GC data, but TiCDC cannot be replicated. This is indeed a quality problem, right?

It was caused by the processing speed of puller’s consumer is far less than the processing speed of data producer of puller, which will lead to data accumulation in puller. It is an quality issue, absolutely  @IANTHEREAL

**Comment by lonng** [comment]
Seems this is a tracking issue, so I will remove the `severity/critical` label and I think the label should attach to sub-tasks.

**Comment by overvenus** [comment]
As of 2021/12, OOM has been resolved for most cases, including cases in described in the issue, change to severity/minor
